File: Linear-Regression/code/model-selection-basic.R
## Coffee example (Coffee Quality Institute, 2018) continued.
coffee <- read.csv("./dataset/coffee_arabica.csv")

# Full model with all predictors
mfull <- lm(
  Flavor ~ factor(Processing.Method) + Aroma + Aftertaste +
    Body + Acidity + Balance + Sweetness + Uniformity + Moisture,
  data = coffee
)

# Check model metrics
summary(mfull)$adj.r.squared
AIC(mfull)
BIC(mfull)

# Exhaustive, brute-force search using leaps package
library(leaps)
all_regs <- regsubsets(
  Flavor ~ .,
  data = coffee,
  nvmax = 10,
  nbest = 2^10,
  really.big = TRUE
)
all_regs_summ <- summary(all_regs)

# View model selection results (commented out for brevity)
# all_regs_summ$which
# all_regs_summ$adjr2
# all_regs_summ$bic

# Organize results according to number of variables in model
p <- 10
k <- c(
  rep(1, choose(p, 1)),
  rep(2, choose(p, 2)),
  rep(3, choose(p, 3)),
  rep(4, choose(p, 4)),
  rep(5, choose(p, 5)),
  rep(6, choose(p, 6)),
  rep(7, choose(p, 7)),
  rep(8, choose(p, 8)),
  rep(9, choose(p, 9)),
  rep(10, choose(p, 10))
)

# Plot adjusted R-squared by number of predictors
boxplot(
  all_regs_summ$adjr2 ~ k,
  xlab = "Number of predictors",
  ylab = "Adjusted R-squared",
  ylim = c(0, 1)
)
abline(h = c(0, 1), lty = 2, col = "red")

# Plot BIC by number of predictors
boxplot(all_regs_summ$bic ~ k, 
        xlab = "Number of predictors", 
        ylab = "BIC")

# Find best models by criteria
max(all_regs_summ$adjr2)
bestR2adj <- which.max(all_regs_summ$adjr2)
min(all_regs_summ$bic)
bestBIC <- which.min(all_regs_summ$bic)

# Check which predictors are in the best models
all_regs_summ$which[bestR2adj,]
all_regs_summ$which[bestBIC,]

# Create dummy variables for processing method
coffee$wet <- ifelse(coffee$Processing.Method == 'Washed / Wet', 1, 0) # 1 = wet, 0 otherwise
coffee$semi <- ifelse(coffee$Processing.Method == 'Semi-washed / Semi-pulped', 1, 0) # 1 = semi/dry, 0 otherwise
coffee$Processing.Method <- NULL

# Fit best adjusted R-squared model
m_bestr2adj <- lm(
  Flavor ~ wet + Aroma + Aftertaste +
    Body + Acidity + Balance + Sweetness + Uniformity + Moisture,
  data = coffee
)
summary(m_bestr2adj)
AIC(m_bestr2adj)
BIC(m_bestr2adj)

# Fit best BIC model (more parsimonious)
m_bestBIC <- lm(Flavor ~ wet + Aroma + Aftertaste +
                  Body + Acidity + Sweetness, data = coffee)
summary(m_bestBIC)
AIC(m_bestBIC)
BIC(m_bestBIC)

# Stepwise methods using MASS package
library(MASS)

# Define full and empty models for stepwise selection
full <- lm(Flavor ~ ., data = coffee)
empty <- lm(Flavor ~ 1, data = coffee)

# Forward selection using AIC (default)
m_f_AIC <- stepAIC(
  object = empty,
  scope = list(upper = full, lower = empty),
  direction = "forward",
  trace = 0
)

# Forward selection using BIC (k = log(n))
m_f <- stepAIC(
  object = empty,
  scope = list(upper = full, lower = empty),
  direction = "forward",
  trace = 0,
  k = log(nrow(coffee))
)
summary(m_f)

# Backward elimination using BIC
m_b <- stepAIC(
  object = full,
  scope = list(upper = full, lower = empty),
  direction = "backward",
  trace = 0,
  k = log(nrow(coffee))
)
summary(m_b)

# Hybrid forward-backward selection using BIC
m_h <- stepAIC(
  object = empty,
  scope = list(upper = full, lower = empty),
  direction = "both",
  trace = 0,
  k = log(nrow(coffee))
)
summary(m_h)

# Note: With 10 variables, all 3 stepwise approaches identify 
# the same BIC-based model as the exhaustive search

# It seems we choose the model with 6 predictors 

File: Linear-Regression/code/model-selection-basic.py
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols
import itertools
from math import comb
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
import warnings
warnings.filterwarnings('ignore')

# Load data
coffee = pd.read_csv("./dataset/coffee_arabica.csv")

# Create dummy variables for Processing.Method
coffee['wet'] = (coffee['Processing.Method'] == 'Washed / Wet').astype(int)
coffee['semi'] = (coffee['Processing.Method'] == 'Semi-washed / Semi-pulped').astype(int)
coffee = coffee.drop('Processing.Method', axis=1)

# Full model with all predictors
formula_full = 'Flavor ~ wet + semi + Aroma + Aftertaste + Body + Acidity + Balance + Sweetness + Uniformity + Moisture'
mfull = ols(formula_full, data=coffee).fit()

# Check model metrics
print("Full Model - Adjusted R-squared:", mfull.rsquared_adj)
print("Full Model - AIC:", mfull.aic)
print("Full Model - BIC:", mfull.bic)
print()

# Exhaustive search function
def exhaustive_search(data, target_var, max_vars):
    predictors = [col for col in data.columns if col != target_var]
    n = len(data)
    
    results = []
    models_info = []
    
    # Test all combinations from 1 to max_vars predictors
    for k in range(1, max_vars + 1):
        for combo in itertools.combinations(predictors, k):
            formula = f"{target_var} ~ " + " + ".join(combo)
            try:
                model = ols(formula, data=data).fit()
                results.append({
                    'predictors': combo,
                    'k': k,
                    'adj_r2': model.rsquared_adj,
                    'aic': model.aic,
                    'bic': model.bic,
                    'formula': formula
                })
                models_info.append((combo, model))
            except:
                continue
    
    results_df = pd.DataFrame(results)
    return results_df, models_info

# Perform exhaustive search
all_regs_df, models_info = exhaustive_search(coffee, 'Flavor', 10)

# Organize results for plotting
k_values = all_regs_df['k'].values
adj_r2_values = all_regs_df['adj_r2'].values
bic_values = all_regs_df['bic'].values

# Plot adjusted R-squared by number of predictors
plt.figure(figsize=(12, 5))

plt.subplot(1, 2, 1)
boxplot_data = [adj_r2_values[k_values == i] for i in range(1, 11)]
plt.boxplot(boxplot_data, labels=range(1, 11))
plt.axhline(y=0, linestyle='--', color='red', alpha=0.7)
plt.axhline(y=1, linestyle='--', color='red', alpha=0.7)
plt.xlabel('Number of predictors')
plt.ylabel('Adjusted R-squared')
plt.title('Adjusted R-squared by Number of Predictors')

# Plot BIC by number of predictors
plt.subplot(1, 2, 2)
boxplot_data_bic = [bic_values[k_values == i] for i in range(1, 11)]
plt.boxplot(boxplot_data_bic, labels=range(1, 11))
plt.xlabel('Number of predictors')
plt.ylabel('BIC')
plt.title('BIC by Number of Predictors')

plt.tight_layout()
plt.show()

# Find best models by criteria
best_adjr2_idx = all_regs_df['adj_r2'].idxmax()
best_bic_idx = all_regs_df['bic'].idxmin()

print("Best Adjusted R-squared:", all_regs_df.loc[best_adjr2_idx, 'adj_r2'])
print("Best BIC:", all_regs_df.loc[best_bic_idx, 'bic'])
print()

# Check which predictors are in the best models
print("Best Adj R² model predictors:", all_regs_df.loc[best_adjr2_idx, 'predictors'])
print("Best BIC model predictors:", all_regs_df.loc[best_bic_idx, 'predictors'])
print()

# Fit best adjusted R-squared model
best_adjr2_predictors = list(all_regs_df.loc[best_adjr2_idx, 'predictors'])
formula_best_adjr2 = f"Flavor ~ {' + '.join(best_adjr2_predictors)}"
m_bestr2adj = ols(formula_best_adjr2, data=coffee).fit()

print("Best Adjusted R-squared Model:")
print(m_bestr2adj.summary())
print("AIC:", m_bestr2adj.aic)
print("BIC:", m_bestr2adj.bic)
print()

# Fit best BIC model (more parsimonious)
best_bic_predictors = list(all_regs_df.loc[best_bic_idx, 'predictors'])
formula_best_bic = f"Flavor ~ {' + '.join(best_bic_predictors)}"
m_bestBIC = ols(formula_best_bic, data=coffee).fit()

print("Best BIC Model:")
print(m_bestBIC.summary())
print("AIC:", m_bestBIC.aic)
print("BIC:", m_bestBIC.bic)
print()

# Stepwise selection function
def stepwise_selection(data, target_var, direction='forward', criterion='bic'):
    predictors = [col for col in data.columns if col != target_var]
    n = len(data)
    
    if direction == 'forward':
        included = []
        current_score = float('inf')
        
        while True:
            changed = False
            excluded = list(set(predictors) - set(included))
            best_new_score = current_score
            
            for new_column in excluded:
                formula = f"{target_var} ~ {' + '.join(included + [new_column])}" if included else f"{target_var} ~ {new_column}"
                model = ols(formula, data=data).fit()
                score = model.bic if criterion == 'bic' else model.aic
                
                if score < best_new_score:
                    best_new_score = score
                    best_candidate = new_column
                    changed = True
            
            if changed:
                included.append(best_candidate)
                current_score = best_new_score
            else:
                break
                
        formula = f"{target_var} ~ {' + '.join(included)}"
        return ols(formula, data=data).fit(), included
    
    elif direction == 'backward':
        included = predictors.copy()
        current_score = ols(f"{target_var} ~ {' + '.join(included)}", data=data).fit().bic
        
        while True:
            changed = False
            best_new_score = current_score
            
            for column in included:
                temp_included = included.copy()
                temp_included.remove(column)
                formula = f"{target_var} ~ {' + '.join(temp_included)}"
                model = ols(formula, data=data).fit()
                score = model.bic if criterion == 'bic' else model.aic
                
                if score < best_new_score:
                    best_new_score = score
                    best_remove = column
                    changed = True
            
            if changed:
                included.remove(best_remove)
                current_score = best_new_score
            else:
                break
                
        formula = f"{target_var} ~ {' + '.join(included)}"
        return ols(formula, data=data).fit(), included

# Stepwise methods
print("Forward selection using BIC:")
m_forward, forward_predictors = stepwise_selection(coffee, 'Flavor', direction='forward', criterion='bic')
print(m_forward.summary())
print("Selected predictors:", forward_predictors)
print()

print("Backward elimination using BIC:")
m_backward, backward_predictors = stepwise_selection(coffee, 'Flavor', direction='backward', criterion='bic')
print(m_backward.summary())
print("Selected predictors:", backward_predictors)
print()

# Note: The hybrid/both direction is more complex to implement but 
# the forward/backward results should be consistent with your R output

print("Note: All selection approaches should identify the same optimal model")
print("Based on the analysis, we choose the model with 6 predictors")